src/libraries/System.Collections/src/System/Collections/Generic/PriorityQueue.cs

```diff
@@ -0,0 +1,695 @@
+        /// </summary>
+        /// <returns>
+        /// <see langword="true"/> if the priority queue is non-empty; <see langword="false"/> otherwise.
+        /// </returns>
+        public bool TryPeek([MaybeNullWhen(false)] out TElement element, [MaybeNullWhen(false)] out TPriority priority)
+        {
+            if (_size != 0)
+            {
+                (element, priority) = _nodes[RootIndex];
+                return true;
+            }
+
+            element = default;
+            priority = default;
+            return false;
+        }
+
+        /// <summary>
+        /// Combined enqueue/dequeue operation, generally more efficient than sequential Enqueue/Dequeue calls.
+        /// </summary>
+        public TElement EnqueueDequeue(TElement element, TPriority priority)
+        {
+            if (element is null)
+            {
+                throw new ArgumentNullException(nameof(element));
+            }
+
+            if (priority is null)
+            {
+                throw new ArgumentNullException(nameof(priority));
+            }
+
+            var root = _nodes[RootIndex];
+
+            if (this.Comparer.Compare(priority, root.priority) < 0)
+            {
+                return element;
+            }
+            else
+            {
+                var newRoot = (element, priority);
+                _nodes[RootIndex] = newRoot;
+
+                MoveDown(newRoot, RootIndex);
+                _version++;
+
+                return root.element;
+            }
+        }
+
+        /// <summary>
+        /// Enqueues a collection of element/priority pairs.
+        /// </summary>
+        public void EnqueueRange(IEnumerable<(TElement Element, TPriority Priority)> items)
+        {
+            if (items is null)
+            {
+                throw new ArgumentNullException(nameof(items));
+            }
+
+            foreach (var (element, priority) in items)
+            {
+                Enqueue(element, priority);
+            }
+        }
+
+        /// <summary>
+        /// Enqueues a collection of elements, each associated with the specified priority.
+        /// </summary>
+        public void EnqueueRange(IEnumerable<TElement> elements, TPriority priority)
+        {
+            if (elements is null)
+            {
+                throw new ArgumentNullException(nameof(elements));
+            }
+
+            foreach (var element in elements)
+            {
+                Enqueue(element, priority);
+            }
+        }
+
+        /// <summary>
+        /// Removes all items from the priority queue.
+        /// </summary>
+        public void Clear()
+        {
+            if (RuntimeHelpers.IsReferenceOrContainsReferences<(TElement, TPriority)>())
+            {
+                // Clear the elements so that the gc can reclaim the references
+                Array.Clear(_nodes, 0, _size);
+            }
+            _size = 0;
+            _version++;
```

Comment:
You could avoid incrementing the version if the heap is empty.

Reply:
Hmm, this behavior is consistent with `Stack<T>` though — what do you suggest then?

Reply:
Fair enough then, I suppose there is precedent.